### PR TITLE
Reply adaptability codelab

### PR DESCRIPTION
**Pull request by Gurupreet**
Based out of already approved :
 https://github.com/android/compose-samples/tree/Gurupreet/reply-m3 which is full sample app in progress.
 
 This codelab shows the emphasis on adaptability and accessibility part of the sample.
 
 [Codelab doc](https://docs.google.com/document/d/1WNI_T_8RODkIs_0b-mIWSkGixpoWfXAluJN7pXRE8Pg/edit#)

Screen Shot 2022-04-11 at 20 02 10

### CODE REVIEW

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepositoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.example.reply.data
+
+import com.example.reply.data.local.LocalAccountsDataProvider
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+
+class AccountsRepositoryImpl : AccountsRepository {
```
Comment (ricknout) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepositoryImpl.kt:8]: Do we need a comment indicating this is a mock (hence the `delay`s)? Same with the other repositories

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/WindowStateUtils.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * Opinionated set of viewport breakpoints
+ *     - Compact: Most phones in portrait mode
+ *     - Medium: Most foldables and tablets in portrait mode
+ *     - Expanded: Most tablets in landscape mode
+ *
+ * More info: https://m3.material.io/foundations/adaptive-design/overview
+ */
+enum class WindowSize { COMPACT, MEDIUM, EXPANDED }
```
Comment (ricknout) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/WindowStateUtils.kt:43]: We need a TODO here to use `androidx.compose.material3:material3-window-size-class`:
https://developer.android.com/jetpack/androidx/releases/compose-material3#1.0.0-alpha10
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/material3/material3-window-size-class/

**Review by ricknout — COMMENTED** [review]
The .idea files are all still in here even though they're in the .gitignore. I also think we can get rid of the example tests

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/WindowStateUtils.kt (hunk 2 above)
Comment (Gurupreet) [reply ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/WindowStateUtils.kt:43]: Yes I plan to add it soon.

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepositoryImpl.kt (hunk 1 above)
Comment (Gurupreet) [reply ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepositoryImpl.kt:8]: Sure let me add comments

**Comment by florina-muntenescu** [comment]
Overall:
* Please rename the project to something consistent with the other codelabs e.g. `AdaptiveUICodelab`
* Remove `.idea` folder
* You're missing the copyright at the top of a lot of files (kt and non-kt)
* Readme is missing

Individual comments to be added to each file

**Review by florina-muntenescu — CHANGES_REQUESTED** [review]
Please split this into 2 different projects, one to contain the starting code, to be merged in `main` and one to contain the finished code to be merged in `end`.
Please do not use `start` and `finish` packages under the `ui` folder

**Comment by Gurupreet** [comment]
> Please split this into 2 different projects, one to contain the starting code, to be merged in `main` and one to contain the finished code to be merged in `end`. Please do not use `start` and `finish` packages under the `ui` folder

Oh I wasn't aware I was following one of the codelab: https://github.com/googlecodelabs/android-compose-codelabs/tree/main/ThemingCodelab/app/src/main/java/com/codelab/theming/ui

I will update it setting up separate projects.

**Comment by florina-muntenescu** [comment]
> > Please split this into 2 different projects, one to contain the starting code, to be merged in `main` and one to contain the finished code to be merged in `end`. Please do not use `start` and `finish` packages under the `ui` folder
> 
> Oh I wasn't aware I was following one of the codelab: https://github.com/googlecodelabs/android-compose-codelabs/tree/main/ThemingCodelab/app/src/main/java/com/codelab/theming/ui
> 
> I will update it setting up separate projects.

The Accessibility codelab is our standard.

File: ReplyAdaptabilityCodelab/app/src/androidTest/java/com/example/reply/ExampleInstrumentedTest.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.example.reply
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/androidTest/java/com/example/reply/ExampleInstrumentedTest.kt:1]: delete file

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/Account.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    val uid: Long,
+    val firstName: String,
+    val lastName: String,
+    val email: String,
+    val altEmail: String,
+    @DrawableRes val avatar: Int,
+    var isCurrentAccount: Boolean = false
+) {
+    val fullName: String = "$firstName $lastName"
+
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/Account.kt:36]: remove new line

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepository.kt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.example.reply.data
+
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * An Interface contract to get all accounts info for User.
+ */
+interface AccountsRepository {
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepository.kt:8]: why is this an interface? why not just have the implementation?

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepositoryImpl.kt (hunk 1 above)
Comment (florina-muntenescu) [reply ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/AccountsRepositoryImpl.kt:8]: Just that this file is never used. Let's remove it together with the interface for it.

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/Email.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    val threads: List<Email> = emptyList()
+) {
+    val senderPreview: String = "${sender.fullName} - 4 hrs ago"
+    val hasBody: Boolean = body.isNotBlank()
+    val hasAttachments: Boolean = attachments.isNotEmpty()
+    val recipientsPreview: String = recipients
+        .map { it.firstName }
+        .fold("") { name, acc -> "$acc, $name" }
+    val nonUserAccountRecipients = recipients
+        .filterNot { LocalAccountsDataProvider.isUserAccount(it.uid) }
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/Email.kt:45]: all of these are not used. Remove.

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/Email.kt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+}
+
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/Email.kt:47]: remove extra new line

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/EmailsRepository.kt [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package com.example.reply.data
+
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * An Interface contract to get all enails info for a User.
+ */
+interface EmailsRepository {
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/EmailsRepository.kt:8]: why do we need an interface? Can we just directly use the implementation?

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/EmailsRepositoryImpl.kt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.example.reply.data.local.LocalEmailsDataProvider
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+
+class EmailsRepositoryImpl : EmailsRepository {
+
+    override fun getAllEmails(): Flow<List<Email>> = flow {
+        delay(1500)
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/EmailsRepositoryImpl.kt:11]: remove the delay completely for all functions.
For what you're teaching in the codelab you don't need a delay.

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/WindowStateUtils.kt (hunk 2 above)
Comment (florina-muntenescu) [reply ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/WindowStateUtils.kt:43]: Let's directly use the m3 window-size-classes. I prefer not to merge the codelab without this change.

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/finish/ReplyApp.kt [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+            contentType = ReplyContentType.LIST_AND_DETAIL
+        }
+    }
+
+    val drawerState = rememberDrawerState(initialValue = DrawerValue.Closed)
+    val scope = rememberCoroutineScope()
+
+    val selectedDestination = ReplyDestinations.INBOX
+
+    if (navigationType == ReplyNavigationType.PERMANENT_NAVIGATION_DRAWER) {
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/finish/ReplyApp.kt:98]: can we extract the navigation to a different function? right now the composable is quite long and hard to read

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/finish/ReplyListContent.kt [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    email: Email,
+    modifier: Modifier = Modifier.padding(horizontal = 16.dp, vertical = 4.dp)) {
+    Card(modifier = modifier, containerColor = MaterialTheme.colorScheme.surface) {
+        Column(
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(20.dp)
+        ) {
+            Row(modifier = Modifier.fillMaxWidth()) {
+                com.example.reply.ui.start.ReplyProfileImage(
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/finish/ReplyListContent.kt:142]: import

File: ReplyAdaptabilityCodelab/build.gradle [hunk 12]
```diff
@@ -0,0 +1,17 @@
+buildscript {
+    ext {
+        kotlin_version = '1.6.10'
+        compose_version = '1.1.1'
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/build.gradle:4]: use compose 1.2

**Review by florina-muntenescu — CHANGES_REQUESTED** [review]
Extra comment: do we have the copyright for all of the images added here?

**Comment by florina-muntenescu** [comment]
I tried running the finished app but it crashes with
```
Empty list doesn't contain element at index 0.
```

on `(ReplyListContent.kt:60)`: `items(replyHomeUIState.emails[selectedItemIndex].threads) { email ->`

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/ReplyHomeViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,38 @@
+package com.example.reply.ui
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.reply.data.Email
+import com.example.reply.data.EmailsRepository
+import com.example.reply.data.EmailsRepositoryImpl
+import kotlinx.coroutines.flow.*
```
Comment (florina-muntenescu) [inline ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/ui/ReplyHomeViewModel.kt:8]: don't use `*` imports

File: ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/EmailsRepository.kt (hunk 8 above)
Comment (Gurupreet) [reply ReplyAdaptabilityCodelab/app/src/main/java/com/example/reply/data/EmailsRepository.kt:8]: We can skip the interfaces, I was just following the contract structure of Interface + Implementation for this sample.

File: AdaptiveUICodelab/app/src/test/java/com/example/reply/ExampleUnitTest.kt [hunk 14]
```diff
@@ -0,0 +1,16 @@
+package com.example.reply
```
Comment (florina-muntenescu) [inline AdaptiveUICodelab/app/src/test/java/com/example/reply/ExampleUnitTest.kt:1]: remove

File: AdaptiveUICodelab/app/src/main/res/drawable/ic_launcher_background.xml [hunk 15]
```diff
@@ -0,0 +1,170 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (florina-muntenescu) [inline AdaptiveUICodelab/app/src/main/res/drawable/ic_launcher_background.xml:1]: copyright missing

**Review by florina-muntenescu — COMMENTED** [review]
Almost there.
Multiple files miss the copyright.

File: AdaptiveUICodelab/gradlew.bat [hunk 16]
```diff
@@ -0,0 +1,89 @@
+@rem
+@rem Copyright 2015 the original author or authors.
```
[the remark this answers is not in the document]
Comment (Gurupreet) [reply AdaptiveUICodelab/gradlew.bat:2]: Yes auto generated and is same for all codelabs

**Review by florina-muntenescu — APPROVED** [review]

**Pull request merged**